src/Networking/IK.cpp

```diff
@@ -1,27 +1,70 @@
 
+#include "IK.h"
 #include "ParseBaseStation.h"
 #include "motor_interface.h"
 #include "../Constants.h"
+#include "../Globals.h"
+#include "../log.h"
 #include <iostream>
 
-// TODO check with electronics to see if this is accurate
-double RAD_PER_INT = (2*M_PI) / ((double) (1<<16));
+// Angular positions from the motor boards are given in millidegrees
+double RAD_PER_INT = (2*M_PI) / (360*1000.0);
 
-int32_t radToInt(double d, double offset, int32_t sign_flip)
+// See IK.h for explanation of sign / offset conventions
+//                                     arm_base, shoulder,                elbow
+const std::array<int, 3> sign_flips = {1,        1,                       1};
+const std::array<double, 3> offsets = {0.0,      Constants::SHOULDER_MAX, Constants::ELBOW_MAX};
+
+int32_t radToInt(double d, int motor_serial)
 {
+  if (!motorSupportsPID(motor_serial))
+  {
+    log(LOG_ERROR, "Invalid motor serial %d for radToInt\n", motor_serial);
+    return 0;
+  }
+  double offset = offsets[motor_serial-1];
+  int sign_flip = sign_flips[motor_serial-1];
   int32_t i = sign_flip * (d - offset) / RAD_PER_INT;
   return i;
 }
 
-double intToRad(int32_t i, double offset, int32_t sign_flip)
+double intToRad(int32_t i, int motor_serial)
 {
+  if (!motorSupportsPID(motor_serial))
+  {
+    log(LOG_ERROR, "Invalid motor serial %d for intToRad\n", motor_serial);
+    return 0;
+  }
+  double offset = offsets[motor_serial-1];
+  int sign_flip = sign_flips[motor_serial-1];
   double d = sign_flip * i * RAD_PER_INT + offset;
   return d;
 }
 
+std::array<double, 3> forward_kinematics() {
+  // This function assumes we already received encoder data from each motor
+  int arm_base_pos = Globals::status_data["arm_base"]["angular_position"];
+  int shoulder_pos = Globals::status_data["shoulder"]["angular_position"];
+  int elbow_pos = Globals::status_data["elbow"]["angular_position"];
+  // See IK.h for explanation of sign / offset conventions
+  double arm_base = intToRad(arm_base_pos, 1);
+  double shoulder = intToRad(shoulder_pos, 2);
+  double elbow = intToRad(elbow_pos, 3);
+  log(LOG_DEBUG, "Running forward kinematics for %f %f %f\n", arm_base, shoulder, elbow);
+  double r1 = Constants::SHOULDER_LENGTH * cos(shoulder);
+  double r2 = Constants::ELBOW_LENGTH * cos(shoulder-elbow);
+  double z1 = Constants::SHOULDER_LENGTH * sin(shoulder);
+  double z2 = Constants::ELBOW_LENGTH * sin(shoulder-elbow);
+  double z = z1+z2;
+  double r = r1+r2;
+  double x = r * cos(arm_base);
+  double y = r * sin(arm_base);
+  return {x,y,z};
+}
+
 bool ParseIKPacket(json &message) {
-  double ELBOW_LENGTH = Constants::ELBOW_LENGTH;
-  double SHOULDER_LENGTH = Constants::SHOULDER_LENGTH;
+  const double ELBOW_LENGTH = Constants::ELBOW_LENGTH;
+  const double SHOULDER_LENGTH = Constants::SHOULDER_LENGTH;
```

Comment:
Is this to make the code less wordy? Also might want to make these constexpr
```suggestion
  constexpr double ELBOW_LENGTH = Constants::ELBOW_LENGTH;
  constexpr double SHOULDER_LENGTH = Constants::SHOULDER_LENGTH;
```

Reply:
Yeah, the values are squared below so it seemed worth it to save some line length. Probably not the cleanest way though. Thanks for the suggestions!